### ISSUE TITLE
SYCL program reports (possibly misdetected) failed OpenCL link on a Vega 56 using ROCm + Ubuntu 20.04

### ISSUE THREAD

**Issue by c0d3st0rm** [intel/llvm#1671]
This issue is currently outlined [here](https://github.com/RadeonOpenCompute/ROCm/issues/1108) in the ROCm repository. I have a program which uses sycl to offload compute, and i can very it works as expected on the host, and both the CPU (when using the Intel compute runtime) and an Intel iGPU using the Neo OpenCL drivers from the 20.04 repos. However, the program fails to run on a Vega 56 using ROCm 3.3.0 on Ubuntu 20.04 (which is currently an unsupported platform as far as I know). The trace is as follows:

```
...

amd_comgr_do_action:
	  ActionKind: AMD_COMGR_ACTION_LINK_RELOCATABLE_TO_EXECUTABLE
	     IsaName: amdgcn-amd-amdhsa--gfx900
	     Options: ""
	        Path: 
	    Language: AMD_COMGR_LANGUAGE_NONE
COMGR::InProcessDriver::Execute argv: lld "/tmp/comgr-3efff3/input/linked.bc.o" "-shared" "-o" "/tmp/comgr-3efff3/output/a.so"
	ReturnStatus: AMD_COMGR_STATUS_SUCCESS

OpenCL API failed. OpenCL API returns: -17 (CL_LINK_PROGRAM_FAILURE) -17 (CL_LINK_PROGRAM_FAILURE)
```

The last line is outputted by dpc++ (one of the daily builds). What's going on here? The link doesn't seem to have failed, judging by comgr's output.

Any help will greatly be appreciated.

**Comment by bader** [comment]
Most likely ROCm compiler is not able to build SPIR-V file produced by the SYCL compiler.
It might be either some unsupported SPIR-V extension or just a bug in ROCm JIT compiler.

**Comment by c0d3st0rm** [comment]
Is the produced SPIR-V saved anywhere as an intermediary?

**Comment by bader** [comment]
You can save SPIR-V by setting `SYCL_DUMP_IMAGES` environment variable.

**Comment by c0d3st0rm** [comment]
So I think I traced it down to:

https://github.com/intel/llvm/blob/c22e34b39ad2ab38f885b2313a4f5a22dd7b0553/sycl/source/detail/program_manager/program_manager.cpp#L572

`LibFileName = "libsycl-fallback-cassert.spv"`;

Is it failing to load a fallback library?

**Comment by bader** [comment]
Did you use https://github.com/intel/llvm/blob/sycl/buildbot/configure.py for "fresh compile"?
You might need to remove CMakeCache.txt and add `devicelib` project to the build (it's done by the configure script automatically).

BTW, I recall that AMD driver doesn't support SPIR-V. See similar issue: https://github.com/intel/llvm/issues/1059#issuecomment-582453655

**Comment by c0d3st0rm** [comment]
No, I'll try the script and get back to you when I have something.

I would be disappointed if that were true, although I wouldn't be surprised. I might have to look at alternative hardware at this rate.
